plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/data/ExecuteInsertBatchImpl.java

```diff
@@ -159,54 +166,77 @@ StringBuilder prepareQueryForStatement(
+        }
+
         StringBuilder valuesPart = new StringBuilder(64);
-        valuesPart.append("(");
-        hasKey = false;
         boolean skipBindValues = CommonUtils.toBoolean(options.get(DBSDataManipulator.OPTION_SKIP_BIND_VALUES));
-        for (int i = 0; i < attributes.length; i++) {
-            DBSAttributeBase attribute = attributes[i];
-            if (DBUtils.isPseudoAttribute(attribute) || (!useMultiRowInsert && (!allNulls && DBUtils.isNullValue(attributeValues[i])))) {
-                continue;
+        for (int i = 0; i < filteredAttributeValues.length; i++) {
+            int j = i % filteredAttributes.length;
+
+            if (j == 0) {
+                valuesPart.append("(");
             }
-            if (hasKey) valuesPart.append(","); //$NON-NLS-1$
-            hasKey = true;
+            DBSAttributeBase attribute = filteredAttributes[j];
+            DBDValueHandler valueHandler = filteredHandlers[i % filteredHandlers.length];
 
-            DBDValueHandler valueHandler = handlers[i];
-            
             if (valueHandler instanceof DBDValueBinder) {
-                valuesPart.append(((DBDValueBinder) valueHandler).makeQueryBind(attribute, attributeValues[i]));
+                valuesPart.append(((DBDValueBinder) valueHandler).makeQueryBind(attribute, filteredAttributeValues[i]));
             } else {
-                valuesPart.append("?"); //$NON-NLS-1$
-            }
-        }
-        valuesPart.append(")"); //$NON-NLS-1$
-        if (useMultiRowInsert) {
-            for (int i = 0; i < attributeValues.length / attributes.length; i++) {
-                if (i != 0) {
-                    query.append(",");
+                if (skipBindValues) {
+                    valuesPart.append(SQLUtils.convertValueToSQL(session.getDataSource(), attribute, valueHandler, filteredAttributeValues[i], DBDDisplayFormat.NATIVE));
+                } else {
+                    valuesPart.append("?"); //$NON-NLS-1$
                 }
-                query.append(valuesPart);
             }
-        } else {
-            query.append(valuesPart);
-        }
-
-        if (skipBindValues) {
-            // Here we replace all binding marks (?) with real values
-            int curPos = 0;
-            for (int i = 0; i < attributeValues.length; i++) {
-                String sqlValue = SQLUtils.convertValueToSQL(
-                    session.getDataSource(),
-                    attributes[i % attributes.length],
-                    handlers[i % handlers.length],
-                    attributeValues[i],
-                    DBDDisplayFormat.NATIVE);
-
-                int qmPos = query.indexOf("?", curPos);
-                query.replace(qmPos, qmPos + 1, sqlValue);
+            if (j == filteredAttributes.length - 1) {
+                valuesPart.append(")"); //$NON-NLS-1$
+                if (i < filteredAttributeValues.length - 1) {
+                    valuesPart.append(","); //$NON-NLS-1$
```

Comment:
Let's use `StringJoiner` for the purpose of joining values with comma. It can also help appending `(` and `)` manually.

Reply:
Fixed it to remove all extra variables except usedAttributes and used StringJoiner. Hope this looks good. I don't feel bad all the changes at all as long as the import functionality works better.